File: sdk/js/src/device.ts
// Copyright 2026 Ojima Abraham
// SPDX-License-Identifier: Apache-2.0

/// GPU detection for the WAVE JavaScript/TypeScript SDK.

import { execSync } from "child_process";
import { platform } from "os";

/** GPU vendor classification. */
export type GpuVendor = "apple" | "nvidia" | "amd" | "intel" | "emulator";

/** Detected GPU device information. */
export interface DeviceInfo {
  vendor: GpuVendor;
  name: string;
}

/** Detect the best available GPU device. */
export async function device(): Promise<DeviceInfo> {
  if (platform() === "darwin") {
    return { vendor: "apple", name: "Apple GPU (Metal)" };
  }

  try {
    const out = execSync("nvidia-smi --query-gpu=name --format=csv,noheader", {
      encoding: "utf-8",
      stdio: ["pipe", "pipe", "pipe"],
    });
    const name = out.trim().split("\n")[0];
    return { vendor: "nvidia", name: `${name} (CUDA)` };
  } catch {
    // nvidia-smi not available
  }

  try {
    const out = execSync("rocminfo", {
      encoding: "utf-8",
      stdio: ["pipe", "pipe", "pipe"],
    });
    if (out.includes("gfx")) {
      return { vendor: "amd", name: "AMD GPU (ROCm)" };
    }
  } catch {
    // rocminfo not available
  }

  try {
    const out = execSync("sycl-ls", {
      encoding: "utf-8",
      stdio: ["pipe", "pipe", "pipe"],
    });
    if (out.includes("level_zero:gpu") || out.includes("opencl:gpu")) {
      return { vendor: "intel", name: "Intel GPU (SYCL)" };
    }
  } catch {
    // sycl-ls not available
  }

  return { vendor: "emulator", name: "WAVE Emulator (no GPU)" };
}

File: sdk/python/src/wave_gpu/device.py
# Copyright 2026 Ojima Abraham
# SPDX-License-Identifier: Apache-2.0

"""GPU detection for the WAVE Python SDK."""

import platform
import subprocess
from typing import Tuple


class DeviceInfo:
    """Detected GPU device information."""

    def __init__(self, vendor: str, name: str) -> None:
        self.vendor: str = vendor
        self.name: str = name

    def __repr__(self) -> str:
        return self.name


def detect_gpu() -> Tuple[str, str]:
    """Detect the best available GPU.

    Returns a (vendor, name) tuple. Vendor is one of: 'apple', 'nvidia',
    'amd', 'intel', 'emulator'.
    """
    if platform.system() == "Darwin":
        return ("apple", "Apple GPU (Metal)")

    try:
        result = subprocess.run(
            ["nvidia-smi", "--query-gpu=name", "--format=csv,noheader"],
            capture_output=True,
            check=True,
            text=True,
        )
        name = result.stdout.strip().split("\n")[0]
        return ("nvidia", f"{name} (CUDA)")
    except (FileNotFoundError, subprocess.CalledProcessError):
        pass

    try:
        result = subprocess.run(
            ["rocminfo"], capture_output=True, check=True, text=True
        )
        if "gfx" in result.stdout:
            for line in result.stdout.splitlines():
                if "Marketing Name" in line:
                    name = line.split(":", 1)[1].strip()
                    return ("amd", f"{name} (ROCm)")
            return ("amd", "AMD GPU (ROCm)")
    except (FileNotFoundError, subprocess.CalledProcessError):
        pass

    try:
        result = subprocess.run(["sycl-ls"], capture_output=True, check=True, text=True)
        if "level_zero:gpu" in result.stdout or "opencl:gpu" in result.stdout:
            return ("intel", "Intel GPU (SYCL)")
    except (FileNotFoundError, subprocess.CalledProcessError):
        pass

    return ("emulator", "WAVE Emulator (no GPU)")


def device() -> DeviceInfo:
    """Detect and return the best available GPU device."""
    vendor, name = detect_gpu()
    return DeviceInfo(vendor, name)
